### PR TITLE
[Dwarf] Support `__ptrauth` qualifier in metadata nodes

### PR DESCRIPTION
**Pull request by kovdan01**
Reland #82363 after fixing build failure https://lab.llvm.org/buildbot/#/builders/5/builds/41428.

Memory sanitizer detects usage of `RawData` union member which is not
filled directly. Instead, the code relies on filling `Data` union
member, which is a struct consisting of signing schema parameters.

According to https://en.cppreference.com/w/cpp/language/union, this is UB:
"It is undefined behavior to read from the member of the union that
wasn't most recently written".

Instead of relying on compiler allowing us to do dirty things, do not
use union and only store `RawData`. Particular ptrauth parameters are
obtained on demand via bit operations.

Original PR description below.

Emit `__ptrauth`-qualified types as `DIDerivedType` metadata nodes in IR with tag `DW_TAG_LLVM_ptrauth_type`, baseType referring to the type which has the qualifier applied, and the following parameters representing the signing schema:

- `ptrAuthKey` (integer)
- `ptrAuthIsAddressDiscriminated` (boolean)
- `ptrAuthExtraDiscriminator` (integer)
- `ptrAuthIsaPointer` (boolean)
- `ptrAuthAuthenticatesNullValues` (boolean)

Co-authored-by: Ahmed Bougacha <ahmed@bougacha.org>

### CODE REVIEW

File: llvm/include/llvm/IR/DebugInfoMetadata.h [hunk 1]
```diff
@@ -975,34 +975,29 @@ class DIDerivedType : public DIType {
...
     PtrAuthData(unsigned Key, bool IsDiscr, unsigned Discriminator,
                 bool IsaPointer, bool AuthenticatesNullValues) {
       assert(Key < 16);
       assert(Discriminator <= 0xffff);
-      Payload.Data.Key = Key;
-      Payload.Data.IsAddressDiscriminated = IsDiscr;
-      Payload.Data.ExtraDiscriminator = Discriminator;
-      Payload.Data.IsaPointer = IsaPointer;
-      Payload.Data.AuthenticatesNullValues = AuthenticatesNullValues;
+      RawData = (Key << 0) | (IsDiscr ? (1 << 4) : 0) | (Discriminator << 5) |
```
Comment (asl) [inline llvm/include/llvm/IR/DebugInfoMetadata.h:985]: Please add comment describing the layout of `RawData`.

File: llvm/include/llvm/IR/DebugInfoMetadata.h [hunk 2]
```diff
@@ -972,6 +972,35 @@ class DIStringType : public DIType {
...
+                (IsaPointer ? (1 << 21) : 0) |
+                (AuthenticatesNullValues ? (1 << 22) : 0);
+    }
+    bool operator==(struct PtrAuthData Other) const {
+      return RawData == Other.RawData;
+    }
+    bool operator!=(struct PtrAuthData Other) const {
+      return !(*this == Other);
+    }
+
```
Comment (dwblaikie) [inline llvm/include/llvm/IR/DebugInfoMetadata.h:995]: Prefer non-member operator overloads (they can be friends and defined inline if that's useful/more convenient than defining them outside the class definition) to allow equal conversions on LHS and RHS.

File: llvm/include/llvm/IR/DebugInfoMetadata.h [hunk 3]
```diff
@@ -972,6 +972,35 @@ class DIStringType : public DIType {
...
+    PtrAuthData(unsigned FromRawData) : RawData(FromRawData) {}
+    PtrAuthData(unsigned Key, bool IsDiscr, unsigned Discriminator,
+                bool IsaPointer, bool AuthenticatesNullValues) {
+      assert(Key < 16);
+      assert(Discriminator <= 0xffff);
+      RawData = (Key << 0) | (IsDiscr ? (1 << 4) : 0) | (Discriminator << 5) |
+                (IsaPointer ? (1 << 21) : 0) |
+                (AuthenticatesNullValues ? (1 << 22) : 0);
+    }
+    bool operator==(struct PtrAuthData Other) const {
```
Comment (dwblaikie) [inline llvm/include/llvm/IR/DebugInfoMetadata.h:989]: Drop the `struct` qualifier as it's not needed (well, it's needed if the name is overloaded to be a struct and something else, but probably best not to do that) in C++.

File: llvm/include/llvm/IR/DebugInfoMetadata.h [hunk 4]
```diff
@@ -1048,6 +1088,39 @@ class DIDerivedType : public DIType {
     return DWARFAddressSpace;
   }
 
+  std::optional<PtrAuthData> getPtrAuthData() const;
+
+  /// \returns The PointerAuth key.
+  std::optional<unsigned> getPtrAuthKey() const {
```
Comment (dwblaikie) [inline llvm/include/llvm/IR/DebugInfoMetadata.h:1094]: Are these helpers particularly useful? It seems like they change the usage code from this:
```
if (auto PtrAuthData = getPtrAuthData())
  do_something_with(PtrAuthData->Key());
```
to
```
if (auto Key = getPtrAuthKey())
  do_something_with(*Key);
```

Which doesn't seem like a huge win?

Comment (kovdan01) [reply llvm/include/llvm/IR/DebugInfoMetadata.h]: These are just small shortcuts which do not give any value except what you described above. I personally have no strong preference, so accepting your suggestion - deleted the helpers. See f765247a3d49b3b726a7442d8d31c1cbd4782154

File: llvm/include/llvm/IR/DebugInfoMetadata.h (hunk 3 above)
Comment (kovdan01) [reply llvm/include/llvm/IR/DebugInfoMetadata.h]: Fixed, thanks, see f765247a3d49b3b726a7442d8d31c1cbd4782154

File: llvm/include/llvm/IR/DebugInfoMetadata.h (hunk 2 above)
Comment (kovdan01) [reply llvm/include/llvm/IR/DebugInfoMetadata.h]: Fixed, thanks, see f765247a3d49b3b726a7442d8d31c1cbd4782154

File: llvm/include/llvm/IR/DebugInfoMetadata.h (hunk 1 above)
Comment (kovdan01) [reply llvm/include/llvm/IR/DebugInfoMetadata.h:991]: Fixed, thanks, see f765247a3d49b3b726a7442d8d31c1cbd4782154

**Comment by kovdan01** [comment]
Would be glad to see @JDevlieghere's and anyone else's interested feedback on the changes

File: llvm/include/llvm/IR/DebugInfoMetadata.h [hunk 5]
```diff
@@ -972,6 +972,35 @@ class DIStringType : public DIType {
...
+                bool IsaPointer, bool AuthenticatesNullValues) {
+      assert(Key < 16);
+      assert(Discriminator <= 0xffff);
+      RawData = (Key << 0) | (IsDiscr ? (1 << 4) : 0) | (Discriminator << 5) |
+                (IsaPointer ? (1 << 21) : 0) |
+                (AuthenticatesNullValues ? (1 << 22) : 0);
+    }
+
+    unsigned Key() { return (RawData >> 0) & 0b1111; }
+    bool IsAddressDiscriminated() { return (RawData >> 4) & 1; }
```
Comment (MaskRay) [inline llvm/include/llvm/IR/DebugInfoMetadata.h:997]: Nit: `functionName`. The code base is unfortunately inconsistent, but newer functions mostly stick with the recommended convention.

File: llvm/lib/AsmParser/LLParser.cpp [hunk 6]
```diff
@@ -5145,19 +5149,30 @@ bool LLParser::parseDIDerivedType(MDNode *&Result, bool IsDistinct) {
...
+  OPTIONAL(ptrAuthAuthenticatesNullValues, MDBoolField, );
   PARSE_MD_FIELDS();
 #undef VISIT_MD_FIELDS
 
   std::optional<unsigned> DWARFAddressSpace;
   if (dwarfAddressSpace.Val != UINT32_MAX)
     DWARFAddressSpace = dwarfAddressSpace.Val;
+  std::optional<DIDerivedType::PtrAuthData> PtrAuthData;
+  if (ptrAuthKey.Val)
+    PtrAuthData = DIDerivedType::PtrAuthData(
```
Comment (MaskRay) [inline llvm/lib/AsmParser/LLParser.cpp:5166]: Nit: `emplace` https://en.cppreference.com/w/cpp/utility/optional/emplace

File: llvm/test/DebugInfo/AArch64/ptrauth.ll [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+; CHECK-NEXT: DW_AT_type {{.*}}"void *"
+; CHECK-NEXT: DW_AT_LLVM_ptrauth_key (0x04)
+; CHECK-NEXT: DW_AT_LLVM_ptrauth_address_discriminated (true)
+; CHECK-NEXT: DW_AT_LLVM_ptrauth_extra_discriminator (0x04d6)
+; CHECK-NEXT: DW_AT_LLVM_ptrauth_isa_pointer	(true)
+; CHECK-NEXT: DW_AT_LLVM_ptrauth_authenticates_null_values	(true)
+
+target datalayout = "e-m:o-i64:64-i128:128-n32:64-S128"
+
+@p = common global i8* null, align 8, !dbg !0
```
Comment (MaskRay) [inline llvm/test/DebugInfo/AArch64/ptrauth.ll:44]: `i8*` indicates a typed pointer, which LLVM is migrating away. Use `ptr`.
Consider a non-common linkage. COMMON linkage is also obsoleted.

File: llvm/lib/IR/DIBuilder.cpp [hunk 8]
```diff
@@ -296,7 +296,20 @@ DIStringType *DIBuilder::createStringType(StringRef Name,
...
+}
+
+DIDerivedType *DIBuilder::createPtrAuthQualifiedType(
+    DIType *FromTy, unsigned Key, bool IsAddressDiscriminated,
+    unsigned ExtraDiscriminator, bool IsaPointer,
+    bool AuthenticatesNullValues) {
+  return DIDerivedType::get(VMContext, dwarf::DW_TAG_LLVM_ptrauth_type, "",
+                            nullptr, 0, nullptr, FromTy, 0, 0, 0, std::nullopt,
+                            std::optional<DIDerivedType::PtrAuthData>(
+                                std::in_place, Key, IsAddressDiscriminated,
```
Comment (kovdan01) [inline llvm/lib/IR/DIBuilder.cpp:309]: According to https://llvm.org/docs/CodingStandards.html#c-standard-versions, we can rely on C++17, so use `std::in_place` for in-place initialization of `std::optional` (see (6) https://en.cppreference.com/w/cpp/utility/optional/optional). Please let me know if we for some reason can't rely on this C++17 feature here.

File: llvm/test/DebugInfo/AArch64/ptrauth.ll (hunk 7 above)
Comment (kovdan01) [reply llvm/test/DebugInfo/AArch64/ptrauth.ll]: Fixed, thanks, see 1adcade6f8d54d8e8ebebc52709e4b293450d117. Regarding linkage - any type is applicable for the test, so just removed that, and by default that should be `external`.

File: llvm/lib/AsmParser/LLParser.cpp (hunk 6 above)
Comment (kovdan01) [reply llvm/lib/AsmParser/LLParser.cpp]: Fixed, thanks, see 1adcade6f8d54d8e8ebebc52709e4b293450d117

File: llvm/include/llvm/IR/DebugInfoMetadata.h (hunk 5 above)
Comment (kovdan01) [reply llvm/include/llvm/IR/DebugInfoMetadata.h]: Fixed, thanks, see 1adcade6f8d54d8e8ebebc52709e4b293450d117

**Comment by kovdan01** [comment]
Would be glad to see @JDevlieghere's and anyone else's interested feedback on the changes.

@MaskRay I've addressed your comments in 1adcade6f8d54d8e8ebebc52709e4b293450d117 - would be glad if you let me know if the fixes are OK.

**Review by JDevlieghere — APPROVED** [review]
I would've probably done the inverse and shifted the `RawData` in the constructor, stored them as individual fields in a struct and have the getters operate on that. And of course have a new vector to return the raw data. But it looks like we rely on `RawData` more than on the getters, so this might be slightly more efficient. LGTM!

File: llvm/lib/Bitcode/Reader/MetadataLoader.cpp [hunk 9]
```diff
@@ -1566,8 +1566,17 @@ Error MetadataLoader::MetadataLoaderImpl::parseOneMetadata(
...
     Metadata *Annotations = nullptr;
-    if (Record.size() > 13 && Record[13])
-      Annotations = getMDOrNull(Record[13]);
+    std::optional<DIDerivedType::PtrAuthData> PtrAuthData;
+
+    // Only look for annotations/ptrauth if both are allocated.
+    // If not, we can't tell which was intended to be embedded, as both ptrauth
+    // and annotations have been expected at Record[13] at various times.
+    if (Record.size() > 14) {
+      if (Record[13])
```
Comment (adrian-prantl) [inline llvm/lib/Bitcode/Reader/MetadataLoader.cpp:1575]: This if  is redundant?

**Review by adrian-prantl — APPROVED** [review]

**Comment by MaskRay** [comment]
> Would be glad to see @JDevlieghere's and anyone else's interested feedback on the changes.
> 
> @MaskRay I've addressed your comments in [1adcade](https://github.com/llvm/llvm-project/commit/1adcade6f8d54d8e8ebebc52709e4b293450d117) - would be glad if you let me know if the fixes are OK.

Thanks. I'm trying to study this patch as I am not so familiar with debug info IR.

If you don't mind waiting a bit, waiting for @dwblaikie (another debug information code owner) could be useful.

**Comment by MaskRay** [comment]
I was confused as I haven't attended any [LLVM Pointer Authentication sync-ups](https://discourse.llvm.org/t/llvm-pointer-authentication-sync-ups/62661) and I am definitely lacking context. It seems that there is an upstreaming proposal back in 2019 https://discourse.llvm.org/t/rfc-pointer-authentication-for-arm64e/53433 and apple/llvm-project contains these `DW_TAG_LLVM_ptrauth_type` changes. This PR is part of the upstream efforts. Yeah, looks good to me.

Normally, when I see `createPtrAuthQualifiedType` with no in-tree user, I'd ask that you prepare links for the immediate user.

**Review by MaskRay — APPROVED** [review]

**Comment by JDevlieghere** [comment]
@dwblaikie Did you want to have a look at this before we merge this? This PR is going to cause quite a bit of conflicts downstream so I would be nice to merge this somewhat early in the week.

**Comment by dwblaikie** [comment]
> @dwblaikie Did you want to have a look at this before we merge this? This PR is going to cause quite a bit of conflicts downstream so I would be nice to merge this somewhat early in the week.

*glances* What sort of conflicts/breakages are you expecting? I guess API breaks on DIBuilder for custom frontends? Other things to look out for?

**Comment by JDevlieghere** [comment]
> What sort of conflicts/breakages are you expecting? I guess API breaks on DIBuilder for custom frontends? Other things to look out for?

Just regular merge conflicts when we merge this into the Swift fork. This PR is upstreaming ptrauth support and started off based on the downstream code, but after a few iterations of review it looks a bit different. That's totally fine, as the result is better for it, but git will need a hand with the merge. The only reason I'm bringing it up is because last time this PR landed on a Friday afternoon (before it got revert) which left us little to deal with all this.

**Comment by dwblaikie** [comment]
> > What sort of conflicts/breakages are you expecting? I guess API breaks on DIBuilder for custom frontends? Other things to look out for?
> 
> Just regular merge conflicts when we merge this into the Swift fork. This PR is upstreaming ptrauth support and started off based on the downstream code, but after a few iterations of review it looks a bit different. That's totally fine, as the result is better for it, but git will need a hand with the merge. The only reason I'm bringing it up is because last time this PR landed on a Friday afternoon (before it got revert) which left us little to deal with all this.

Ah, your own downstream, fair enough. It looks right enough to me, FWIW :)

**Review by dwblaikie — APPROVED** [review]

**Pull request merged**